src/storage-client/src/controller/collection_mgmt.rs

```diff
@@ -53,26 +53,46 @@ impl CollectionManager {
         write_handle: persist_handles::PersistWriteWorker<T>,
         now: NowFn,
     ) -> CollectionManager {
-        let collections = Arc::new(Mutex::new(BTreeSet::new()));
+        let collections: Arc<Mutex<BTreeSet<GlobalId>>> = Arc::new(Mutex::new(BTreeSet::new()));
         let collections_outer = Arc::clone(&collections);
+
         // Note(parkmycar): The capacity here was chosen randomly.
         let (tx, mut rx) = mpsc::channel::<(
             GlobalId,
             Vec<(Row, Diff)>,
             oneshot::Sender<Result<(), StorageError>>,
         )>(256);
 
+        // Allows callers to wait until we finish any in-progress work.
+        //
+        // For example, after unregistering a collection a user might wait on a barrier to be sure
+        // any in-progress work with that collection has completed.
+        let barrier: Arc<Mutex<Vec<oneshot::Sender<_>>>> = Arc::new(Mutex::new(Vec::new()));
+        let barrier_outer = Arc::clone(&barrier);
+
         mz_ore::task::spawn(|| "ControllerManagedCollectionWriter", async move {
             let mut interval = tokio::time::interval(tokio::time::Duration::from_millis(1_000));
             loop {
+                // Notify any waiting barriers.
+                barrier
+                    .lock()
+                    .expect("CollectionManager panicked")
+                    .drain(..)
+                    .for_each(|waiter| {
+                        let _ = waiter.send(());
+                    });
+
                 tokio::select! {
                     _ = interval.tick() => {
-                        let collections = &mut *collections.lock().await;
-
-                        let now = T::from(now());
-                        let updates = collections.iter().map(|id| {
-                            (*id, vec![], now.clone())
-                        }).collect::<Vec<_>>();
+                        // Update each collection.
+                        let updates = {
```

Comment:
Just noting here that this is an example of the shortening of how long the lock is held that I am apprehensive about.